test/Base_Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/base_test_helpers/ExecutableLocation.java

```diff
@@ -0,0 +1,72 @@
+package org.enso.base_test_helpers;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.module.ModuleFinder;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.util.Arrays;
+import org.graalvm.nativeimage.ImageInfo;
+
+public final class ExecutableLocation {
+  private ExecutableLocation() {}
+
+  /** Returns canonical path to the enso executable location. */
+  public static String getExecutableLocation() {
+    File exec;
+    if (ImageInfo.inImageRuntimeCode()) {
+      exec = getExecutableLocationInNI();
+    } else {
+      exec = getExecutableLocationFromModule();
+    }
+    try {
+      return exec.getCanonicalPath();
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  private static File getExecutableLocationInNI() {
+    var codeSource = ExecutableLocation.class.getProtectionDomain().getCodeSource();
+    assert codeSource != null;
+    var loc = codeSource.getLocation();
+    try {
+      return new File(loc.toURI());
+    } catch (URISyntaxException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  private static File getExecutableLocationFromModule() {
+    var mpStr = System.getProperty("jdk.module.path");
```

Comment:
can it be null?

Reply:
No. If the application does not run in NI, but in JVM, it must have been started with proper module path, and with `org.enso.runner` module. Otherwise, it would not even boot.

Reply:
Certain properties are treated as _guaranteed to be set_ by many libraries and applications. I know it can then cause problems with alternatives VMs like [Bck2Brwsr VM](https://github.com/jtulach/bck2brwsr/issues/28). But yeah, in this case ...

> Otherwise, it would not even boot.

...we can expect the property to be set. Especially because we know what JVM we are running on.